src/services/project/validate.rs

```diff
@@ -0,0 +1,66 @@
+use crate::db::models::{ProjectInsert, ProjectRequest};
+use crate::db::schema::{events, projects};
+use crate::utils::Result;
+use diesel::dsl::*;
+use diesel::prelude::*;
+use diesel::sql_types;
+use phf::phf_map;
+use slug::slugify;
+
+#[derive(Debug, Fail, PartialEq)]
+pub enum ValidateError {
+    #[fail(display = "Project name cannot end in '-': {}", project_name)]
+    DashSuffix { project_name: String },
+    #[fail(display = "Cannot name project {}, is a reserved word", project_name)]
+    ReservedWord { project_name: String },
+}
+
+static RESERVED_WORDS: phf::Map<&'static str, ()> = phf_map! {
+    "admin" => (),
+    "login" => (),
+    "submit" => (),
+    "delete" => (),
+};
+
+sql_function!(fn nextval(a: sql_types::Text) -> sql_types::BigInt);
+
+pub fn call(project: ProjectRequest, conn: &diesel::PgConnection) -> Result<ProjectInsert> {
```

Comment:
How do you think we should integrate the service module architecture into the auth refactors? i.e. what should the control flow look like? Right now I'm thinking something like this potentially:

1. We have a `Authorizable` trait (not attached to that name) and we have a function on `DatabaseContext` like this:
    ```rust
    pub fn get_connection(&self, auth: AuthorizationInfo) -> T where T: Authorizable {
    // Body here
    }
    ```
2. `resolvers` are responsible for knowing which `validate` to call and which service to call once validated. `DatabaseContext.validate` takes information from the resolver and calls a service for validation, which implements the `Authorizable` trait for some database type, and then `DatabaseContext.validate` returns the raw connection (or some kind of "read-only" connection) to the calling resolver, which then calls the service necessary to return information.